ethereum/rpc/namespaces/debug/api.go

```diff
@@ -112,6 +113,100 @@ func (a *API) TraceTransaction(hash common.Hash, config *evmtypes.TraceConfig) (
 	return decodedResult, nil
 }
 
+// TraceBlockByNumber returns the structured logs created during the execution of
+// EVM and returns them as a JSON object.
+func (a *API) TraceBlockByNumber(height rpctypes.BlockNumber, config *evmtypes.TraceConfig) ([]*evmtypes.TxTraceResult, error) {
+	a.logger.Debug("debug_traceBlockByNumber", "height", height)
+	if height == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+	// Get Tendermint Block
+	resBlock, err := a.backend.GetTendermintBlockByNumber(height)
+	if err != nil {
+		return nil, err
+	}
+
+	return a.traceBlock(height, config, resBlock.Block.Txs)
+}
+
+// traceBlock configures a new tracer according to the provided configuration, and
+// executes all the transactions contained within. The return value will be one item
+// per transaction, dependent on the requested tracer.
+func (a API) traceBlock(height rpctypes.BlockNumber, config *evmtypes.TraceConfig, txs types.Txs) ([]*evmtypes.TxTraceResult, error) {
+	txsLength := len(txs)
+
+	if txsLength == 0 {
+		// If there are no transactions return empty array
+		return []*evmtypes.TxTraceResult{}, nil
+	}
+
+	var (
+		results = make([]*evmtypes.TxTraceResult, txsLength)
+		wg = new(sync.WaitGroup)
+		jobs = make(chan *evmtypes.TxTraceTask, txsLength)
+	)
+
+	threads := runtime.NumCPU()
+	if threads > txsLength {
+		threads = txsLength
+	}
+	wg.Add(threads)
+	for th := 0; th < threads; th++ {
+		go func() {
+			defer wg.Done()
+			txDecoder := a.clientCtx.TxConfig.TxDecoder()
+			// Fetch and execute the next transaction trace tasks
+			for task := range jobs {
+				tx, err := txDecoder(txs[task.Index])
+				if err != nil {
+					a.logger.Error("failed to decode transaction", "hash", txs[task.Index].Hash(), "error", err.Error())
+					continue
+				}
+
+				messages := tx.GetMsgs()
+				if len(messages) == 0 {
+					continue
+				}
+				ethMessage, ok := messages[0].(*evmtypes.MsgEthereumTx)
+				if !ok {
+					// Just considers Ethereum transactions
+					continue
+				}
+
+				traceTxRequest := &evmtypes.QueryTraceTxRequest{
+					Msg:     ethMessage,
+					TxIndex: uint32(task.Index),
+					TraceConfig: config,
+				}
+
+				res, err := a.queryClient.TraceTx(rpctypes.ContextWithHeight(int64(height)), traceTxRequest)
```

Comment:
you can define `rpctypes.ContextWithHeight(int64(height))` outside the go routine and the for loop since the height is always the same